Merge XPEvo fixed/changes into main repository

Chris92de opened the pull request:
This pull request includes all the changes that were made to the XPEvo branch of EvoSC#, without the additions of ContactAdminModule, ToornamentModule and ServerSyncModule.
Changes to the required dependencies (namely NATS and toornamentapi) have been reverted.

New modules: RoundRankingModule
Updated modules: ScoreboardModule, GameModeUiModule

Full list of changes compared to `master` branch can be seen here: https://github.com/EvoEsports/EvoSC-sharp/compare/master...EvoEsports:EvoSC-sharp-XPEvo:master - keep in mind that some of these changes already have been merged into `master` branch here and were merged into the forked repository after the fact.

Abbreviated list of changes since August 20th:
- Updated design for `CurrentMapModule` by @araszka 
- Updated design and functionality for `SpectatorTargetInfoModule` by @araszka 
- Added `SpectatorCamModeModule` by @araszka - replaces Nadeo's default Spec Cam Mode switcher in the bottom right
- Rename `Scoreboard` to `ScoreboardModule`
- Updated design and functionality for `ScoreboardModule` (incl. themability, nickname support, showing current/max rounds, etc)
- Added `RoundRankingModule` by @araszka - similar functionality as SmallScoresTable in TMWT/Teams_Matchmaking modes
- Added default settings to `GameModeUiModule` for `SmallScoresTable` UILayer
- Few changes to the `Dockerfile` to make sure EvoSC runs as user id `9999` instead of root.

Chris92de commented on docker/docker-compose.yml:
```diff
@@ -19,9 +19,9 @@ services:
+
   evosc_sharp:
-    image: evoscsharp:latest
+    image: evoesports/evosc-sharp:latest
```
Oh I thought this was meant to be an example file for the public, hence why I worded the README.md file the way I did.

Chris92de replied:
This should now be fixed with the latest commit.

snixtho commented:
Because commit history seems to be lost, please create a list of changes in the description.

Chris92de commented:
> Because commit history seems to be lost, please create a list of changes in the description.

As requested, I've updated the description of the pull request with an abbreviated list of changes and also added a link to compare the fork to the current master branch.

Chris92de commented:
The last commit resolves the comment by @AtomicLiquid about the `docker-compose.yml` file not including a build parameter.

snixtho commented on src/Modules/RoundRankingModule/Services/RoundRankingService.cs:
```diff
@@ -0,0 +1,231 @@
+            });
+    }
+
+    public Task HideRoundRankingWidgetAsync() =>
```
This method does not belong in this service.

snixtho commented on src/Modules/RoundRankingModule/Services/RoundRankingService.cs:
```diff
@@ -0,0 +1,231 @@
+namespace EvoSC.Modules.Official.RoundRankingModule.Services;
+
+[Service(LifeStyle = ServiceLifeStyle.Singleton)]
+public class RoundRankingService(
```
After looking at this entire class. I would split this up and only include necessary methods for updating/getting the state variables in this service. And then put methods that does some specialized operation on that data in a transient/scoped service. This way we don't have a do-it-all singleton that is references everywhere in memory and keep the single responsibility principle intact.

By doing this, it is way easier to manage the thread-safe operations on these variables as well.

snixtho commented on src/Modules/RoundRankingModule/Templates/Components/RoundRankingRow.mt:
```diff
@@ -0,0 +1,99 @@
+                    <label
+                            if="checkpoint.GainedPoints == 0"
+                            class="text-primary"
+                            text='{{ checkpoint.IsDNF ? Icons.FlagO : (checkpoint.IsFinish ? Icons.FlagCheckered : checkpoint.CheckpointId + 1) }}'
```
Maybe some sort of utility method would be better here than nesting ternary statements.

Actually would be nice if there the template engine had a way to define C# methods somewhere in the component that can be reused and make the code cleaner.

snixtho commented on src/Modules/RoundRankingModule/Templates/Components/RoundRankingRow.mt:
```diff
@@ -0,0 +1,99 @@
+                        size="{{ width-20 }} {{ height }}"
+                />
+                <label
+                        text="{{ checkpoint.IsDNF ? &quot;DNF&quot; : (checkpoint.TimeDifference!=null ? checkpoint.TimeDifference.ToString().TrimStart('0') : checkpoint.Time.ToString().TrimStart('0') ) }}"
```
Maybe some sort of utility method would be better here than nesting ternary statements.

snixtho commented on src/Modules/ScoreboardModule/Services/ScoreboardNicknamesService.cs:
```diff
@@ -0,0 +1,68 @@
+    IManialinkManager manialinkManager
+) : IScoreboardNicknamesService
+{
+    private readonly Dictionary<string, string> _nicknames = new();
```
missing locks or change to ConcurrentDictionary

snixtho commented on src/Modules/ScoreboardModule/Services/ScoreboardNicknamesService.cs:
```diff
@@ -0,0 +1,68 @@
+        manialinkManager.SendPersistentManialinkAsync("ScoreboardModule.PlayerNicknames",
+            new { nicknames = ToManiaScriptArray(_nicknames) });
+
+    public string ToManiaScriptArray(Dictionary<string, string> nicknameMap)
```
move out of class, this is an utility method

snixtho commented on src/Modules/ScoreboardModule/Services/ScoreboardNicknamesService.cs:
```diff
@@ -0,0 +1,68 @@
+        return $"[{joinedEntries}]";
+    }
+
+    public string ToManiaScriptArrayEntry(KeyValuePair<string, string> loginNickname)
```
utility method, move out of class

snixtho commented on src/Modules/ScoreboardModule/Services/ScoreboardNicknamesService.cs:
```diff
@@ -0,0 +1,68 @@
+        return $"\"{loginNickname.Key}\" => \"{EscapeNickname(loginNickname.Value)}\"";
+    }
+
+    public string EscapeNickname(string nickname)
```
utility method, move out of class

snixtho commented on src/Modules/ScoreboardModule/Services/ScoreboardService.cs:
```diff
@@ -0,0 +1,112 @@
+{
+    private const string ScoreboardTemplate = "ScoreboardModule.Scoreboard";
+    private const string MetaDataTemplate = "ScoreboardModule.MetaData";
+    private readonly object _currentRoundMutex = new();
+    private readonly object _isWarmUpMutex = new();
+    private int _roundNumber = 1;
+    private bool _isWarmUp;
```
Suggestion on what you should do here:
- Move these state variables and related methods out of this class in it's own singleton service.
- Set this service as transient/scoped, depending on how its used.
- Inject the newly-created singleton into this class.

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Body/Legend.mt:
```diff
@@ -0,0 +1,86 @@
+        />
+        <label text="POS"
+               class="text-primary"
+               pos="{{ positionBoxWidth/2.0 }} {{ height / -2 + 0.25 }}"
```
```suggestion
               pos="{{ positionBoxWidth/2f }} {{ height / -2 + 0.25 }}"
```

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Body/Legend.mt:
```diff
@@ -0,0 +1,86 @@
+        />
+        <label text="NAT."
+               class="text-primary"
+               pos="{{ positionBoxWidth+columnSpacing+(flagWidth/2.0) }} {{ height / -2 + 0.25 }}"
```
```suggestion
               pos="{{ positionBoxWidth+columnSpacing+(flagWidth/2f) }} {{ height / -2 + 0.25 }}"
```

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Body/Legend.mt:
```diff
@@ -0,0 +1,86 @@
+        />
+        <label text="CLUB"
+               class="text-primary"
+               pos="{{ positionBoxWidth+columnSpacing+flagWidth+columnSpacing+(clubTagWidth/2.0) }} {{ height / -2 + 0.25 }}"
```
```suggestion
               pos="{{ positionBoxWidth+columnSpacing+flagWidth+columnSpacing+(clubTagWidth/2f) }} {{ height / -2 + 0.25 }}"
```

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Body/Legend.mt:
```diff
@@ -0,0 +1,86 @@
+        <label id="legend_best_time"
+               text="BEST TIME"
+               class="text-primary"
+               pos="{{ width-columnSpacing - 55.0 }} {{ height / -2 + 0.25 }}"
```
```suggestion
               pos="{{ width-columnSpacing - 55f }} {{ height / -2 + 0.25 }}"
```

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Header/HeaderContent.mt:
```diff
@@ -0,0 +1,119 @@
+    <property type="double" name="padding" default="3.4"/>
+
+    <template>
+        <frame id="header_content" pos="{{ width / 2.0 }} {{ height / -2.0 }}">
```
```suggestion
        <frame id="header_content" pos="{{ width / 2f }} {{ height / -2f }}">
```

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Header/HeaderContent.mt:
```diff
@@ -0,0 +1,119 @@
+    <template>
+        <frame id="header_content" pos="{{ width / 2.0 }} {{ height / -2.0 }}">
+            <label id="header_text_left"
+                   pos="{{ float.Parse(Theme.ScoreboardModule_Logo_Width, CultureInfo.InvariantCulture) * -0.5 - 4.0 }} 1.75"
```
```suggestion
                   pos="{{ float.Parse(Theme.ScoreboardModule_Logo_Width, CultureInfo.InvariantCulture) * -0.5 - 4f }} 1.75"
```

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Header/HeaderContent.mt:
```diff
@@ -0,0 +1,119 @@
+        <frame id="header_content" pos="{{ width / 2.0 }} {{ height / -2.0 }}">
+            <label id="header_text_left"
+                   pos="{{ float.Parse(Theme.ScoreboardModule_Logo_Width, CultureInfo.InvariantCulture) * -0.5 - 4.0 }} 1.75"
+                   size="{{ width/2.0-float.Parse(Theme.ScoreboardModule_Logo_Width, CultureInfo.InvariantCulture)-2.0 }} {{ height }}"
```
```suggestion
                   size="{{ width/2f-float.Parse(Theme.ScoreboardModule_Logo_Width, CultureInfo.InvariantCulture)-2f }} {{ height }}"
```

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Header/HeaderContent.mt:
```diff
@@ -0,0 +1,119 @@
+                   valign="center2"
+            />
+            <label id="header_text_left_small"
+                   pos="{{ float.Parse(Theme.ScoreboardModule_Logo_Width, CultureInfo.InvariantCulture) * -0.5 - 4.0 }} -1.75"
```
```suggestion
                   pos="{{ float.Parse(Theme.ScoreboardModule_Logo_Width, CultureInfo.InvariantCulture) * -0.5 - 4f }} -1.75"
```

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Header/HeaderContent.mt:
```diff
@@ -0,0 +1,119 @@
+            />
+            <label id="header_text_left_small"
+                   pos="{{ float.Parse(Theme.ScoreboardModule_Logo_Width, CultureInfo.InvariantCulture) * -0.5 - 4.0 }} -1.75"
+                   size="{{ width/2.0-float.Parse(Theme.ScoreboardModule_Logo_Width, CultureInfo.InvariantCulture)-2.0 }} {{ height }}"
```
```suggestion
                   size="{{ width/2f-float.Parse(Theme.ScoreboardModule_Logo_Width, CultureInfo.InvariantCulture)-2f }} {{ height }}"
```

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Header/HeaderContent.mt:
```diff
@@ -0,0 +1,119 @@
+                   opacity="0.8"
+            />
+            <label id="header_text_right"
+                   pos="{{ float.Parse(Theme.ScoreboardModule_Logo_Width, CultureInfo.InvariantCulture) * 0.5 + 4.0 }} 1.75"
```
```suggestion
                   pos="{{ float.Parse(Theme.ScoreboardModule_Logo_Width, CultureInfo.InvariantCulture) * 0.5 + 4f }} 1.75"
```

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Header/HeaderContent.mt:
```diff
@@ -0,0 +1,119 @@
+            />
+            <label id="header_text_right"
+                   pos="{{ float.Parse(Theme.ScoreboardModule_Logo_Width, CultureInfo.InvariantCulture) * 0.5 + 4.0 }} 1.75"
+                   size="{{ width/2.0-float.Parse(Theme.ScoreboardModule_Logo_Width, CultureInfo.InvariantCulture)-2.0 }} {{ height }}"
```
```suggestion
                   size="{{ width/2f-float.Parse(Theme.ScoreboardModule_Logo_Width, CultureInfo.InvariantCulture)-2f }} {{ height }}"
```

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Header/HeaderContent.mt:
```diff
@@ -0,0 +1,119 @@
+                   valign="center2"
+            />
+            <label id="header_text_right_small"
+                   pos="{{ float.Parse(Theme.ScoreboardModule_Logo_Width, CultureInfo.InvariantCulture) * 0.5 + 4.0 }} -1.75"
```
```suggestion
                   pos="{{ float.Parse(Theme.ScoreboardModule_Logo_Width, CultureInfo.InvariantCulture) * 0.5 + 4f }} -1.75"
```

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Header/HeaderContent.mt:
```diff
@@ -0,0 +1,119 @@
+            />
+            <label id="header_text_right_small"
+                   pos="{{ float.Parse(Theme.ScoreboardModule_Logo_Width, CultureInfo.InvariantCulture) * 0.5 + 4.0 }} -1.75"
+                   size="{{ width/2.0-float.Parse(Theme.ScoreboardModule_Logo_Width, CultureInfo.InvariantCulture)-2.0 }} {{ height }}"
```
```suggestion
                   size="{{ width/2f-float.Parse(Theme.ScoreboardModule_Logo_Width, CultureInfo.InvariantCulture)-2f }} {{ height }}"
```

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Header/HeaderContent.mt:
```diff
@@ -0,0 +1,119 @@
+        </frame>
+    </template>
+
+    <script once="true"><!--
```
Move script into its own file.

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Row/Framemodel.mt:
```diff
@@ -0,0 +1,244 @@
+                                   height="{{ rowHeight }}"
+            />
+
+            <frame pos="{{ positionBoxWidth + columnSpacing }} {{ rowHeight / -2.0 }}" z-index="10">
```
```suggestion
            <frame pos="{{ positionBoxWidth + columnSpacing }} {{ rowHeight / -2f }}" z-index="10">
```

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Row/Framemodel.mt:
```diff
@@ -0,0 +1,244 @@
+
+            <frame pos="{{ positionBoxWidth + columnSpacing }} {{ rowHeight / -2.0 }}" z-index="10">
+                <!-- Flag -->
+                <Flag height="{{ rowInnerHeight }}" />
```
Could this be a core component or is it specialized to the scoreboard only?

araszka replied:
Could be made a core component. I'll keep it in mind when updating the scoreboard for teams mode.

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Row/Framemodel.mt:
```diff
@@ -0,0 +1,244 @@
+                           class="text-primary"
+                           text="CLUB"
+                           pos="{{ rowInnerHeight * 0.75 }} 0"
+                           size="{{ rowInnerHeight * 2.0 }} {{ rowInnerHeight }}"
```
```suggestion
                           size="{{ rowInnerHeight * 2f }} {{ rowInnerHeight }}"
```

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Row/Framemodel.mt:
```diff
@@ -0,0 +1,244 @@
+                           size="{{ rowInnerHeight * 2.0 }} {{ rowInnerHeight }}"
+                           valign="center2"
+                           halign="center"
+                           textsize="{{ Theme.UI_FontSize*2 }}"
```
```suggestion
                           textsize="{{ Theme.UI_FontSize*2f }}"
```

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Row/Framemodel.mt:
```diff
@@ -0,0 +1,244 @@
+                       pos="{{ rowInnerHeight * 3.0 + columnSpacing * 2.0 }} 0"
+                       size="{{ w / 3.0 }} {{ rowHeight }}"
+                       valign="center2"
+                       textsize="{{ Theme.UI_FontSize*2 }}"
```
Use the `text-xl` utility class instead and for the others too.

Ref:
- `text-xs`: `Theme.UI_FontSize*0.8`
- `text-normal`: `Theme.UI_FontSize` (or `text-primary` for default colors + normal size)
- `text-md`: `Theme.UI_FontSize*1.2`
- `text-lg`: `Theme.UI_FontSize*1.5`
- `text-xl`: `Theme.UI_FontSize*2`
- `text-2xl`: `Theme.UI_FontSize*4`
- `text-3xl`: `Theme.UI_FontSize*6`
- `text-4xl`: `Theme.UI_FontSize*8`

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Row/Framemodel.mt:
```diff
@@ -0,0 +1,244 @@
+    </template>
+
+    <script once="true">
+        <!--
```
move script into its own file

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Row/PlayerActions.mt:
```diff
@@ -0,0 +1,58 @@
+                 valign="center"
+                 halign="center"
+                 text="{{ Icons.VideoCamera }}"
+                 textsize="{{ Theme.UI_FontSize*2 }}"
```
use `text-xl` class

snixtho commented on src/Modules/ScoreboardModule/Templates/Components/Row/PositionBox.mt:
```diff
@@ -0,0 +1,58 @@
+                   pos='{{ (width-(Theme.ScoreboardModule_PositionBox_ShowAccent=="True" ? accentBarWidth : 0)) / 2.0 + (Theme.ScoreboardModule_PositionBox_ShowAccent=="True" ? accentBarWidth : 0) }} {{ height / -2.0 + 0.25 }}'
+                   valign="center"
+                   halign="center"
+                   textsize="{{ Theme.UI_FontSize*2 }}"
```
use `text-xl` class

snixtho commented on src/Modules/ScoreboardModule/Templates/Scoreboard.mt:
```diff
@@ -0,0 +1,532 @@
+    </template>
+
+    <script>
+        <!--
```
move script into its own file
